app/services/combined_audio_derivative_creator.rb

```diff
@@ -0,0 +1,163 @@
+      fingerprint: fingerprint,
+      component_metadata: {durations: durations}
+    }
+  end
+
+  def output_file(format)
+    Tempfile.new(['output', ".#{format}"], :encoding => 'binary').path
+  end
+
+  # Use ffprobe to determine the length of an audio file.
+  def duration_of_audio_file_in_seconds(path)
+    options = ['ffprobe', '-v', 'error',
+      '-show_entries', 'format=duration', '-of',
+      'default=noprint_wrappers=1:nokey=1'
+    ] + [ path ]
+    @cmd.run(*options).out.strip
+  end
+
+  def download_components
+    result = []
+    audio_member_files.each do |original_file|
+      log "Downloading #{original_file.metadata['filename']}"
+      new_temp_file = Tempfile.new(['temp_', original_file.metadata['filename'].downcase], :encoding => 'binary')
+      original_file.open(rewindable:false) do |input_audio_io|
+        new_temp_file.write input_audio_io.read until input_audio_io.eof?
+      end
+      result << new_temp_file
+    end
+    result
+  end
+
+  # Generate ffmpeg command to concatenate a set of audio files using
+  # ffmpeg stream mapping.
+  # This is tricky but documented at
+  # https://trac.ffmpeg.org/wiki/Map .
+  # For now, the audio settings are hardwired.
+  # This does not run the command; it just returns
+  # an array that can be passed to @cmd.run .
+  def args_for_ffmpeg(output_file_path)
+    # ffmpeg -y: overwrite the already-existing temp file
+    ffmpeg_command = ['ffmpeg', '-y']
+
+    # Then a list of input files specified with -i
+    input_files = @components.map {|x| [ "-i", x.path] }.flatten
+
+    # List the number of audio streams: one per file
+    stream_list = 0.upto(@components.count - 1).to_a.map{ |n| "[#{n}:a]"}.join
+
+    # Specify what to do with the audio streams:
+    #
+    # Stream mapping:
+    #   Stream #0:0 (mp3float) -> concat:in0:a0
+    #   Stream #1:0 (mp3float) -> concat:in1:a0
+    #   Stream #2:0 (mp3float) -> concat:in2:a0
+    #   [ and so on]
+    #   concat -> Stream #0:0 (libmp3lame)
+    #
+    # v=0 means there is no video.
+    filtergraph = "concat=n=#{@components.count}:v=0:a=1[a]"
+
+    # Finally, some output options:
+    # -map [a] : map just the audio to the output
+    #
+    # -ac 1 : output should be mono
+    #    see https://trac.ffmpeg.org/wiki/AudioChannelManipulation
+    # -b:a 64K: make the output a constant 64k bitrate
+    #    see https://trac.ffmpeg.org/wiki/Encode/MP3
+    output_quality = ["-map", "[a]", "-ac", "1", "-b:a", "64k"]
+
+    # Put it all together and return:
+    ffmpeg_command +
+      input_files +
+      ['-filter_complex'] +
+      ["#{stream_list}#{filtergraph}"] +
+      output_quality +
+      [output_file_path]
+  end
+
+  def audio_member_files
+    audio_members.map { |asset| asset.file }
+  end
+
+  # Filters out non-audio items
+  def audio_members
+    work.members.order(:position, :id).select do |member|
+      member.stored? && member.file.mime_type.start_with?('audio')
+    end
```

Comment:
Cache/memoize this with `@audio_members ||=`, since you refer to it multiple times. Without that, as it is, every time you refer to it, it might go to the db again to fetch it again. Which is a performance problem, but also could be a semantics problem if it's *changed* in the db on subsequent fetch, could lead to unexpected and hard to debug behavior. 

Am I sure this will go to the DB? No, I'm not sure, `work.members` should cache/memoize itself, but I _think_ adding `order` on the end will make it go to the DB. 

But rather than be sure (or be sure if it's called more than once) -- just memoize, and then you know you're safe, even if code changes after you were 'sure'.

Reply:
Yup!